presto-main/src/main/java/com/facebook/presto/sql/gen/BytecodeUtils.java

```diff
@@ -295,14 +321,15 @@ else if (type == ConnectorSession.class) {
         }
         block.append(invoke(binding, name));
 
-        if (function.isNullable()) {
+        if (bestChoice.getReturnPlaceConvention() == ReturnPlaceConvention.STACK && function.isNullable()) {
             block.append(unboxPrimitiveIfNecessary(scope, returnType));
         }
         block.visitLabel(end);
 
-        if (outputBlock.isPresent()) {
+        if (bestChoice.getReturnPlaceConvention() == ReturnPlaceConvention.STACK && outputBlock.isPresent()) {
```

Comment:
Turn this into `switch`. And add comment in the `PROVIDED_BLOCK` branch to describe why it isn't necessary.

I recommend placing `outputBlock.isPresent()` outside the `switch`.